R/survival.R

```diff
@@ -0,0 +1,292 @@
+
+  # Patients with events.
+  df_events <- df_stand[df_stand$status == 1L, , drop = FALSE]
+
+  # Add number of events per unique event time.
+  df_events_per_time <- h_n_events_per_time(df_stand, time = "time", status = "status")
+  df_events <- merge(df_events, df_events_per_time, by = "time", all = TRUE)
+
+  # Calculate the log rank statistic u_l and the variance sigma_l2 iteratively.
+  u_l <- sigma_l2 <- 0
+  for (i in seq_len(nrow(df_events))) {
+    # This event time.
+    t_i <- df_events$time[i]
+
+    # Number of overall events at this time.
+    n_events_ti <- df_events$n_events[i]
+
+    # This treatment arm indicator.
+    trt_i <- df_events$treatment[i]
+
+    # Proportions of patients at risk at time t_i, per arm.
+
+    # Note: Also here we need to use sum divided by n, otherwise
+    # we will get wrong results when there are ties.
+    y_bar_1_ti <- sum(df_stand$treatment & df_stand$time >= t_i) / n
+    y_bar_0_ti <- sum(!df_stand$treatment & df_stand$time >= t_i) / n
+
+    # Adjusted proportion of patients at risk in the treatment arm:
+    y_bar_1_ti_coxph <- y_bar_1_ti * exp(theta)
+
+    # Increment u_l.
+    u_l <- u_l + (trt_i - y_bar_1_ti_coxph / (y_bar_1_ti_coxph + y_bar_0_ti))
+
+    # Increment sigma_l2.
+    y_bar_all_ti <- y_bar_1_ti_coxph + y_bar_0_ti
+
+    sigma_l2_factor_ti <- if (!use_ties_factor || n_events_ti == 1) {
+      1
+    } else {
+      (n * y_bar_all_ti - n_events_ti) / (n * y_bar_all_ti - 1)
+    }
+    sigma_l2_summand <- y_bar_1_ti_coxph * y_bar_0_ti * sigma_l2_factor_ti / y_bar_all_ti^2
+    sigma_l2 <- sigma_l2 + sigma_l2_summand
+  }
+  u_l <- u_l / n
+  sigma_l2 <- sigma_l2 / n
+  se_theta_l <- sqrt(1 / (n * sigma_l2))
+  structure(
+    u_l,
+    sigma_l2 = sigma_l2,
+    se_theta_l = se_theta_l,
+    n = n
+  )
+}
+
+#' Estimate Log Hazard Ratio via Score Function
+#'
+#' This function estimates the log hazard ratio by finding the root of the log-rank score function.
+#'
+#' @details This deactivates the ties factor correction in the score function by passing
+#' `use_ties_factor = FALSE` to the `score_fun`.
+#'
+#' @param score_fun (`function`) The log-rank score function to be used for estimation.
+#' @param interval (`numeric`) A numeric vector of length 2 specifying the interval in which to search for the root.
+#' @param ... Additional arguments passed to `score_fun`.
+#' @return A list containing:
+#' - `theta`: The estimated log hazard ratio.
+#' - `se`: The standard error of the estimated log hazard ratio.
+#' - `sigma_L2`: The variance of the log-rank statistic.
+#' - `n`: The number of observations used in the calculation.
+#'
+#' @keywords internal
+h_log_hr_est_via_score <- function(score_fun, interval = c(-10, 10), ...) {
+  assert_function(score_fun, args = c("theta", "use_ties_factor"))
+  assert_numeric(interval, len = 2L, finite = TRUE)
+  assert_true(interval[1] < interval[2])
+
+  score_solution <- stats::uniroot(
```

Comment:
some assertion might be needed to ensure that the root can be obtained, like the values are of opposite sign in these interval if the function is monotonic (should be)

Reply:
or add some trycatch here?